solidity/test/Oracle_test.js

```diff
@@ -146,43 +146,111 @@ contract('Oracle', () => {
-      let req = await requestDataFrom(oc, link, 0, args)
-      internalId = req.receipt.logs[2].topics[1]
-    })
+    context('successful consumer', () => {
+      beforeEach(async () => {
+        mock = await deploy('examples/GetterSetter.sol')
+        let fHash = functionSelector('requestedBytes32(bytes32,bytes32)')
+        let args = requestDataBytes(specId, mock.address, fHash, requestId, '')
+        let req = await requestDataFrom(oc, link, 0, args)
+        internalId = req.receipt.logs[2].topics[1]
+      })
 
-    context('when called by a non-owner', () => {
-      it('raises an error', async () => {
-        await assertActionThrows(async () => {
-          await oc.fulfillData(internalId, 'Hello World!', {from: stranger})
+      context('when called by a non-owner', () => {
+        it('raises an error', async () => {
+          await assertActionThrows(async () => {
+            await oc.fulfillData(internalId, 'Hello World!', {from: stranger})
+          })
         })
       })
-    })
 
-    context('when called by an owner', () => {
-      it('raises an error if the request ID does not exist', async () => {
-        await assertActionThrows(async () => {
-          await oc.fulfillData(0xdeadbeef, 'Hello World!', {from: oracleNode})
+      context('when called by an owner', () => {
+        it('raises an error if the request ID does not exist', async () => {
+          await assertActionThrows(async () => {
+            await oc.fulfillData(0xdeadbeef, 'Hello World!', {from: oracleNode})
+          })
+        })
+
+        it('sets the value on the requested contract', async () => {
+          await oc.fulfillData(internalId, 'Hello World!', {from: oracleNode})
+
+          let mockRequestId = await mock.requestId.call()
+          assert.equal(requestId.toString(), web3.toUtf8(mockRequestId))
+
+          let currentValue = await mock.getBytes32.call()
+          assert.equal('Hello World!', web3.toUtf8(currentValue))
+        })
+
+        it('does not allow a request to be fulfilled twice', async () => {
+          await oc.fulfillData(internalId, 'First message!', {from: oracleNode})
+          await assertActionThrows(async () => {
+            await oc.fulfillData(internalId, 'Second message!!', {from: oracleNode})
+          })
         })
       })
+    })
+
+    context('malicious consumer', () => {
+      const paymentAmount = toWei(1)
 
-      it('sets the value on the requested contract', async () => {
-        await oc.fulfillData(internalId, 'Hello World!', {from: oracleNode})
+      context('fails during fulfillment', () => {
+        beforeEach(async () => {
+          mock = await deploy('examples/MaliciousConsumer.sol', link.address, oc.address)
+          let fHash = functionSelector('assertFail(bytes32,bytes32)')
+          let args = requestDataBytes(specId, mock.address, fHash, requestId, '')
+          let req = await requestDataFrom(oc, link, paymentAmount, args)
```

Comment:
In the failure scenario, we go through the malicious contract:

```
const req = await mock.requestData('cancelRequestOnFulfill(bytes32,bytes32)')
```

Is there a reason we're bypassing the normal request flow? Is it for better test isolation?

We could have just done
```
const req = await mock.requestData('assertFail(bytes32,bytes32)')
```

Reply:
@dimroc do you mind expanding on what you mean by `we're bypassing the normal request flow?`

`assertFail` & `cancelRequestOnFulfill` are 2 separate failure scenarios.

- `assertFail` would have caused an error in the `Oracle` contract because of the old call to `require(...)` on [L111](https://github.com/smartcontractkit/chainlink/pull/358/files#diff-dc8d0c05dfc85ef9202cbe5f1590ea83L111)
- `cancelRequestOnFulfill` starts to drain the `withdrawableWei` would fail in a game of musical chairs when the last person tries to withdraw

Reply:
Talked about this at standup. In the assertFail tests, we make the request directly to the contract. In the cancelRequestOnFulfill tests, we use the malicious contract as the requester. We only updated the cancel scenario because we needed to, but it'd be good to keep the two consistent if only to clarify there's no real difference in the set up.